src/net_processing.cpp

```diff
@@ -2284,23 +2289,30 @@ void PeerManagerImpl::ProcessGetBlockData(CNode& pfrom, Peer& peer, const CInv&
         if (!(pindex->nStatus & BLOCK_HAVE_DATA)) {
             return;
         }
+        can_direct_fetch = CanDirectFetch();
+        block_pos = pindex->GetBlockPos();
+    }
+
+    const CNetMsgMaker msgMaker(pfrom.GetCommonVersion());
     std::shared_ptr<const CBlock> pblock;
     if (a_recent_block && a_recent_block->GetHash() == pindex->GetBlockHash()) {
         pblock = a_recent_block;
     } else if (inv.IsMsgWitnessBlk()) {
         // Fast-path: in this case it is possible to serve the block directly from disk,
         // as the network format matches the format on disk
         std::vector<uint8_t> block_data;
-        if (!m_chainman.m_blockman.ReadRawBlockFromDisk(block_data, pindex->GetBlockPos())) {
-            assert(!"cannot load block from disk");
+        if (!m_chainman.m_blockman.ReadRawBlockFromDisk(block_data, block_pos)) {
+            LogPrint(BCLog::NET, "Cannot load block from disk. It was likely pruned before we could read it.\n");
+            return;
         }
         m_connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::BLOCK, Span{block_data}));
         // Don't set pblock as we've sent the block
     } else {
         // Send block from disk
         std::shared_ptr<CBlock> pblockRead = std::make_shared<CBlock>();
-        if (!m_chainman.m_blockman.ReadBlockFromDisk(*pblockRead, *pindex)) {
-            assert(!"cannot load block from disk");
+        if (!m_chainman.m_blockman.ReadBlockFromDisk(*pblockRead, block_pos)) {
+            LogPrint(BCLog::NET, "Cannot load block from disk. It was likely pruned before we could read it.\n");
+            return;
```

Comment:
> Last commit: Not sure. This looks like a p2p protocol change?

For the local node, this change means that the node will continue functional instead of crashing. 
Which I think is good. The node shouldn't crash because it received a p2p message that cannot answer.

For the remote peer, this means that the connection will remain active for ~10 more minutes until the block request times out and the peer closes the socket.
Which isn't the best, but it is how we currently handle unanswered block requests.

Are you seeing something else?

Reply:
> > For the local node, this change means that the node will continue functional instead of crashing.
> 
> The node wouldn't crash here, because cs_main was being held, so no pruning could happen in-between.

Could also be that the requested block data is not on disk or not accessible for the time this happen. Something external to our software.

Still, removing the external cause from the equation. The only diff here is that we allow a tiny extra pruning window in-between the index lookup and the block data read.
Because, we currently return early if the block is pruned [here](https://github.com/bitcoin/bitcoin/blob/e789b30b2565c8bdbf48a45f2c5a7b92e5d61d25/src/net_processing.cpp#L2285), without logging anything. And after this PR, we could either return early on the no logging line (if the block is already pruned) or log "Cannot load block from disk. It was likely pruned before we could read it" if the block was pruned right after the index lookup.

Other than that, I'm not seeing any p2p protocol change.

Maybe @andrewtoth could take this to the PR description?

Reply:
> If your datadir was corrupted from outside, I think a crash is acceptable. You likely wouldn't be able to re-org or otherwise stay in consensus anyway.

You know; if the block in conflict is old enough, then the node should be "mostly" ok. The problem would be on new peers performing IBD, which would disconnect from the node for the lack of response. 

> > I'm not seeing any p2p protocol change.
> 
> Maybe "protocol change" was the wrong word? I guess "behavior change" may be better?

Sounds good.

Reply:
If a block is pruned before we lock cs_main, this function returns early. This behavior is not changed in this PR.

If a block is not pruned before we lock cs_main, previously we are guaranteed to send the block to the peer.

This PR changes this to have a short window after unlocking cs_main but before opening the fd to read the block that if the block is pruned and unlinked from another thread, then this function also returns early.

I think we all agree this is not a p2p protocol change? I also believe I documented this behavior change in the PR description and in the debug logs.

Reply:
So the other behavior change is that if we fail to read a block for reasons other than being pruned, previously the node would crash. Now we will log the failure and continue.

> If your datadir was corrupted from outside, I think a crash is acceptable. You likely wouldn't be able to re-org or otherwise stay in consensus anyway.

I suppose if the read fails, we can retake cs_main and check if the block was pruned out from under us. If so, we log and continue, otherwise we can keep crashing on an assert? Do we really want to keep crashing here if we fail to read a block? I'm not sure.

Reply:
> if the read fails, we can retake cs_main and check if the block was pruned out from under us. If so, we log and continue, otherwise we can keep crashing on an assert

@maflcko I updated the code to do this. Thus I don't think there is any behavior change.
If the block requested is pruned, we return early before and after this PR. If the block requested is unpruned and we fail to read it, we will crash before and after this PR.

Reply:
>  I am not asking to add back the assert.

+1 on not re-adding the assert.

> However, I wonder what the point is of forcing the remote peer into a 10 minute delay+disconnect. Why not disconnect immediately? Also, if the remote has set a permission flag to avoid the disconnect (not sure if this exists), I wonder what the correct behavior would be.

If the remote peer has a permission flag to avoid the disconnection, maybe it could respond with a `NOTFOUND` message? Same as we do with tx getdata requests.
But maybe better to discuss this in a follow-up or in an issue?

Reply:
Updated to log and return instead of hitting an assert. Also updated the commit description to clarify this behavior.

> However, I wonder what the point is of forcing the remote peer into a 10 minute delay+disconnect. Why not disconnect immediately?

We could, but I think this is very unlikely to occur in practice. We already disconnect above on line 2293 if we are past the NODE_NETWORK_LIMITED_MIN_BLOCKS threshold, which is MIN_BLOCKS_TO_KEEP (288) so we won't prune below that anyways.